tools/astarte_device_fleet_simulator/lib/astarte_device_fleet_simulator/device.ex

```diff
@@ -0,0 +1,255 @@
+    with {:ok, device_pid} <- fetch_device_pid(realm, device_id),
+         {:ok, [interface]} <- fetch_interface_names(),
+         # blocking
+         :ok <-
+           Device.wait_for_connection(device_pid) do
+      {:ok, _} = Registry.register(AstarteDeviceFleetSimulator.Registry, "device", [])
+
+      new_data =
+        Map.merge(data, %{
+          interface: interface,
+          publication_interval_ms: Config.publication_interval_ms!()
+        })
+
+      if Registry.count(AstarteDeviceFleetSimulator.Registry) == Config.device_count!() do
+        GenServer.cast(Scheduler, :device_spawn_end)
+      end
+
+      if not data.skip_waiting do
+        # wait for all other devices before running
+        Logger.info("#{data.device_id} ready to start sending messages", tag: "device_ready")
+        {:next_state, :waiting, new_data}
+      else
+        # do not wait for all devices
+        Logger.info("#{data.device_id} started sending messages", tag: "device_sending_started")
+        [path: path, value: value, qos: qos] = Config.message_opts()
+
+        {:next_state, :running, new_data,
+         publish_action(Config.publication_interval_ms!(), path, value, qos)}
+      end
+    end
+  end
+
+  def waiting(:cast, :begin_publishing, data) do
+    Logger.info("#{data.device_id} started sending messages", tag: "device_sending_started")
+
+    [path: path, value: value, qos: qos] = Config.message_opts()
+
+    {:next_state, :running, data,
+     publish_action(Config.publication_interval_ms!(), path, value, qos)}
+  end
+
+  def running({:timeout, :publish}, {path, value, qos}, data) do
+    {:ok, device_pid} = fetch_device_pid(data.realm, data.device_id)
+    Task.start(fn -> send_datastream(device_pid, data.interface, path, value, qos) end)
+    {:keep_state_and_data, publish_action(data.publication_interval_ms, path, value, qos)}
+  end
+
+  defp do_register(retry_times, data) do
+    with {:ok, %{body: %{"data" => %{"credentials_secret" => credentials_secret}}}} <-
+           Agent.register_device(data.client, data.device_id),
+         {:ok, _} <-
+           Astarte.Device.start_link(
+             Config.device_opts() ++
+               [device_id: data.device_id, credentials_secret: credentials_secret]
+           ) do
+      {:keep_state,
+       %{
+         client: data.client,
+         device_id: data.device_id,
+         credentials_secret: credentials_secret,
+         realm: data.realm,
+         interface_provider: data.interface_provider,
+         path: data.path,
+         skip_waiting: data.skip_waiting
+       }, connecting_action()}
+    else
+      {:error, reason} ->
+        Logger.error(inspect(reason))
+        backoff_time = compute_backoff_time(retry_times)
+
+        new_retry_times =
+          if retry_times < @max_backoff_exponent, do: retry_times + 1, else: retry_times
+
+        backoff_action = register_action(backoff_time, new_retry_times)
+        {:keep_state_and_data, backoff_action}
```

Comment:
Always use parentheses when _calling_ functions, even with no arguments. `mix format` should fix this.

Reply:
The only function being called here is `register_action(backoff_time, new_retry_times)`, which has parentheses

Reply:
My bad, thought `backoff_action` was a function call